File: app/src/main/java/com/bitflaker/lucidsourcekit/database/notifications/daos/NotificationMessageDao.java
package com.bitflaker.lucidsourcekit.database.notifications.daos;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import com.bitflaker.lucidsourcekit.database.notifications.entities.NotificationMessage;
import com.bitflaker.lucidsourcekit.general.NotificationObfuscationLookup;

import java.util.List;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Single;

@Dao
public interface NotificationMessageDao {
    @Query("SELECT * FROM NotificationMessage ORDER BY notificationCategoryId, obfuscationTypeId, id")
    Single<List<NotificationMessage>> getAll();

    @Query("SELECT * FROM NotificationMessage WHERE notificationCategoryId = :notificationCategoryId ORDER BY obfuscationTypeId, id")
    Single<List<NotificationMessage>> getAllOfCategory(String notificationCategoryId);

    @Query("SELECT * FROM NotificationMessage WHERE notificationCategoryId = :notificationCategoryId AND obfuscationTypeId = :obfuscationTypeId ORDER BY notificationCategoryId, id")
    Single<List<NotificationMessage>> getAllOfCategoryAndObfuscationType(String notificationCategoryId, int obfuscationTypeId);

    @Query("SELECT COUNT(*) FROM NotificationMessage WHERE notificationCategoryId = :notificationCategoryId")
    Single<Integer> getCountOfMessagesForCategory(String notificationCategoryId);

    @Query("SELECT COUNT(*) FROM NotificationMessage WHERE notificationCategoryId = :notificationCategoryId AND obfuscationTypeId = :obfuscationTypeId")
    Single<Integer> getCountOfMessagesForCategoryAndObfuscationType(String notificationCategoryId, int obfuscationTypeId);

    @Query("SELECT notificationCategoryId, obfuscationTypeId, COUNT(*) AS messageCount FROM NotificationMessage GROUP BY notificationCategoryId, obfuscationTypeId")
    Single<List<NotificationObfuscationLookup.NotificationObfuscationCount>> getMessageCountsForObfuscationType();

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    Completable insertAll(List<NotificationMessage> notificationMessages);

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    long insert(NotificationMessage notificationMessage);

    @Delete
    Completable delete(NotificationMessage notificationMessage);

    @Delete
    void deleteAll(List<NotificationMessage> notificationMessages);

    @Update
    Completable update(NotificationMessage notificationMessage);
}

File: app/src/main/java/com/bitflaker/lucidsourcekit/database/notifications/daos/NotificationMessageDao.kt
package com.bitflaker.lucidsourcekit.database.notifications.daos

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Update
import com.bitflaker.lucidsourcekit.database.notifications.entities.NotificationMessage

@Dao
interface NotificationMessageDao {
    @Query("SELECT * FROM NotificationMessage ORDER BY notificationCategoryId, id")
    suspend fun getAll(): List<NotificationMessage>

    @Query("SELECT * FROM NotificationMessage WHERE notificationCategoryId = :notificationCategoryId ORDER BY notificationCategoryId, id")
    suspend fun getAllOfCategory(notificationCategoryId: String): MutableList<NotificationMessage>

    @Query("SELECT COUNT(*) FROM NotificationMessage WHERE notificationCategoryId = :notificationCategoryId")
    suspend fun getCountOfMessagesForCategory(notificationCategoryId: String): Int

    @Query("SELECT COUNT(*) FROM NotificationMessage WHERE notificationCategoryId = :notificationCategoryId")
    suspend fun getCountOfMessagesForCategoryAndObfuscationType(notificationCategoryId: String): Int

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    suspend fun insertAll(notificationMessages: List<NotificationMessage>)

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    suspend fun insert(notificationMessage: NotificationMessage): Long

    @Delete
    suspend fun delete(notificationMessage: NotificationMessage)

    @Delete
    suspend fun deleteAll(notificationMessages: List<NotificationMessage>)

    @Update
    suspend fun update(notificationMessage: NotificationMessage)
}
